### DISCUSSION TITLE
REPL: Run single function OR highlighted code

### DISCUSSION THREAD

**Discussion by RustoMCSpit** [zed-industries/zed#23257] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

So you'd highlight a function via highlighting it with your mouse and then hit something like F6 (or maybe right click its name, whatever) and then you'd be prompted in the terminal to input the values of all the arguments and it would show you what the function returned as well as outputted (etc.). Useful debugging feature.

**Comment by jansol** [comment]
How would you pass arguments to the function?

**Comment by RustoMCSpit** [comment]
> How would you pass arguments to the function?

i said 'prompted in the terminal'? not sure whats not being communicated here sorry. 

take the below pseudo code

def coinToss (color, amount, speed):
CODE HERE

the terminal would be prompt me for all parameters (someone else can figure out if we should make the user define the variable type while inputting too)
Insert color: 
Insert amount:
Insert speed:

and it would then show the outputs of all variables and what returned.

if the function is recursive it would display each recursuions values and if the function had args (undefined amount of parameters) the terminal would ask how many you want to put in

or maybe have it just display return output, or have it as an option for the user. general idea

**Comment by SomeoneToIgnore** [comment]
Do you have any examples from other ide/editors that support this feature?

**Comment by jansol** [comment]
> i said 'prompted in the terminal'? not sure whats not being communicated here sorry.

Ah, sorry. Seems I skipped a line while reading.

@SomeoneToIgnore  One similar thing I can immediately think of is the Matlab GUI where you have a panel with an editable real-time listing of variables in global scope, which will be in force when you run code snippets.

**Comment by RustoMCSpit** [comment]
> Do you have any examples from other ide/editors that support this feature?

no, original idea

**Comment by JosephTLyons** [comment]
It seems like this feature would require us to build custom support for every language. For example, for a single Rust function, we would need to hoist it into some compilable project so that we could run it.  It seems like a huge lift to get that to work for one language, let alone many.

**Comment by RustoMCSpit** [comment]
> It seems like this feature would require us to build custom support for every language. For example, for a single Rust function, we would need to hoist it into some compilable project so that we could run it. It seems like a huge lift to get that to work for one language, let alone many.

*not every language is compiled

**Comment by JosephTLyons** [comment]
> > It seems like this feature would require us to build custom support for every language. For example, for a single Rust function, we would need to hoist it into some compilable project so that we could run it. It seems like a huge lift to get that to work for one language, let alone many.
> 
> *not every language is compiled

Right, a compiled language would need to be compiled, an interpreted language would need an interpreter / repl.

**Comment by RustoMCSpit** [comment]
> It seems like a huge lift to get that to work for one language, let alone many.

well if the interpreter can read the whole project, it can read a single function too. compiled languages would cause issue however. am i missing something here

**Comment by mrnugget** [comment]
I think a good first step would be the ability to "pipe" selected text (or a syntactic node) to an external program. We could do that via tasks even. So say you select text, you hit `cmd-shift-p`, use `task: spawn task with stdin` (or whatever) and type in `ruby` (or whatever) and that `ruby` process would receive the given selection as stdin.

**Comment by RustoMCSpit** [comment]
> to an external program. 

as long as it works with the virtual environment and has access to all modules and files thats fine

> So say you select text

well the idea was just for functions but selected texts works too, we should still have the right click a function and have a "run function" option

> type in ruby

by default, it should just use the interpreter of whatever the rest of the file sees. i feel like inputting a different interpreter is pretty niche but still should be considered

> I think a good first step would be the ability to "pipe" selected text (or a syntactic node) to an external program. We could do that via tasks even. So say you select text, you hit `cmd-shift-p`, use `task: spawn task with stdin` (or whatever) and type in `ruby` (or whatever) and that `ruby` process would receive the given selection as stdin.

sounds fine to me but what about handling of recursive functions or functions with an undefined amount of parameters, etc.? do we just show recursion graphs and have a certain string to stop inputting?

**Comment by JosiahParry** [comment]
> Right, a compiled language would need to be compiled, an interpreted language would need an interpreter / repl.

I think this is really what I'm after. I write Rust and want to be able to run `cargo check` with a key stroke and when I write R or python I want to be able to execute code in my repl with `cmd + enter`

**Comment by RustoMCSpit** [comment]
I've changed the title. Now we're going with right click (secondary click) of a function name or selected code.

**Comment by RustoMCSpit** [comment]
> I've changed the title. Now we're going with right click (secondary click) of a function name or selected code.

just to be clear, zed does have a 'run current script in the integrated terminal'  like vscode right?

**Comment by RustoMCSpit** [comment]
> I think this is really what I'm after. I write Rust and want to be able to run `cargo check` with a key stroke and when I write R or python I want to be able to execute code in my repl with `cmd + enter`

rust is trickier by nature as it's compiled, it's doable though if the devs spend time on it. interpreted is much easier.

**Comment by JosiahParry** [comment]
> > I think this is really what I'm after. I write Rust and want to be able to run `cargo check` with a key stroke and when I write R or python I want to be able to execute code in my repl with `cmd + enter`
> 
> rust is trickier by nature as it's compiled, it's doable though if the devs spend time on it. interpreted is much easier.

Using the development tasks is making it easier. Would be great if the tasks can be bound to keystrokes

**Comment by gsamat** [comment]
I want to reiterate that VS Code has [RunSelectedText](https://code.visualstudio.com/docs/terminal/basics#_run-selected-text) feature which is very useful for shell-wrangling (as mentioned in issue #8656  which is closed as duplicate to this). Instead of editing commands directly in Terminal, you edit your shell code in a good editor and ability to run it via single command is super useful. 

In this Issue, however, there is a need for a 'data input interface', which might be additional work. Maybe keep them separate?

**Comment by polymorphicengine** [comment]
sorry to come to this discussion late, but i have one more suggestion:
it would be great if you could not only run selected code, but also just a line or 'block' of code, where a block is defined by several lines without any free lines inbetween. this is very useful in live coding

**Comment by RustoMCSpit** [comment]
> sorry to come to this discussion late

no apologies needed!

**Comment by RustoMCSpit** [comment]
![image](https://github.com/zed-industries/zed/assets/134429563/5997248c-ba9f-4806-a65a-1bfc5739338c)
it could look something like this dropdown

https://youtu.be/X0OylwLzBQw?si=VgtX9gqr-6siTypn&t=387

**Comment by RustoMCSpit** [comment]
> I want to reiterate that VS Code has [RunSelectedText](https://code.visualstudio.com/docs/terminal/basics#_run-selected-text) feature which is very useful for shell-wrangling (as mentioned in issue #8656 which is closed as duplicate to this). Instead of editing commands directly in Terminal, you edit your shell code in a good editor and ability to run it via single command is super useful.
> 
> In this Issue, however, there is a need for a 'data input interface', which might be additional work. Maybe keep them separate?

can you include a screen recording of that?

**Comment by RustoMCSpit** [comment]
> In this Issue, however, there is a need for a 'data input interface', which might be additional work. Maybe keep them separate?

are you sure? they seem related to me

**Comment by ToddMCr** [comment]
For those who want to run highlighted text, this is working well for me in `~/.config/zed/keymap.json`:
```json
{
    "context": "Editor",
    "bindings": {
      "cmd-enter": ["workspace::SendKeystrokes", "cmd-c ctrl-` cmd-v enter"]
    }
  }
  ```
(found from https://github.com/zed-industries/zed/issues/16102#issuecomment-2299607954)

**Comment by mcabbott** [comment]
Thanks, that's great!

Is there any way to make it work with the terminal on the right? I wrote a longer comment about this in #16102, perhaps best to discuss there.

**Comment by ghost** [comment]
need online help

**Comment by RustoMCSpit** [comment]
apparently this is called READ EVALUATE PRINT LOOP and you can do it in Godot https://www.youtube.com/watch?v=4Q46A_UaY_4&t=496s

**Reply by RustoMCSpit** [reply]
REPL

**Reply by jansol** [reply]
https://zed.dev/docs/repl It has also been available in Zed for a good while via Jupyter kernels (although not all Jupyter features are supported)

**Reply by RustoMCSpit** [reply]
should this discussion be discarded or is there a use to it?

**Reply by RustoMCSpit** [reply]
also @jansol, would you mind if i @ you somewhere else on task running for your input?

**Reply by jansol** [reply]
go ahead

**Comment by casperhart** [comment]
I use Zed and VSCode every day for python development. I heavily rely on the VSCode command "Python: Send current selection / line to terminal" when running scripts interactively, and Zed is lacking this feature.

For me, the existing REPL functionality (jupyter / ipykernel) cannot be used, because I will often connect to remote servers with SSH where ipython is not available. I also just dislike ipython, and would much rather have a basic python terminal open.

For running _highlighted_ text, this can technically be accomplished via tasks, but it doesn't work if you want to infer the appropriate class definition, function definition, or statement based on cursor position, like how VSCode does it.

This might be possible to achieve with tasks if there were more functionality for selecting the relevant syntax node and for navigating between nodes. For example, if we could define a task that would:

1. select the current (or following) top-level syntax node (class definition, function definition, or statement)
2. copy and paste in to the terminal and hit 'return'
4. move the cursor to the end of the syntax node

So these issues may be relevant:
https://github.com/zed-industries/zed/issues/5133
https://github.com/zed-industries/zed/issues/10799

Ideally we could implement this with extensions, but they seem very limited. It's hard to add to zed itself, as what constitutes an appropriate syntax node depends on the language, and for compiled languages, REPLs aren't relevant.

**Comment by ddxv** [comment]
I went to try Zed today and this is a blocker for me to even try running print("hello world"). As others mentioned, it is the way I develop, I think it's very popular with Python/Julia/R data workflows?

Honestly, it's the only way I've ever coded and it's surprising to me that this isn't a built in for Zed to have? I don't want to write all my code in the interpreter, that gets confusing very quick. I of course want to use the editor, like a notepad, building out a function that works, and testing it in the interpreter below/totheside.

**Reply by mcabbott** [reply]
For what it's worth, I use Zed this way, with the following in `~/.config/zed/keymap.json`
```
  {
    "context": "Editor",
    "bindings": {
      "cmd-enter": ["workspace::SendKeystrokes", "cmd-c ctrl-` cmd-v ctrl-` down home"]
    }
  },
```
Modified from earlier in this thread to move to the next line, so that `cmd-enter` several times can run a sequence of lines. At least in the Julia REPL, incomplete lines won't run immediately, so you can step through e.g. a multi-line function definition this way. (You can also select several lines first.)

Not quite as convenient as having the editor select the relevant code block for you. And it will replace whatever was in your clipboard.

**Comment by scfurl** [comment]
# SendCode crate — send code from editor to terminal/REPL (addresses many requests here)

I've been building a `send_code` crate for Zed that addresses many of the workflows discussed in this thread. It's modeled after the Sublime Text [SendCode](https://github.com/randy3k/SendCode) plugin and provides native send-to-terminal functionality without relying on the `SendKeystrokes` workaround (which clobbers your clipboard and can't do block expansion).

## What it does

- **`SendCode`** — sends the current selection (or auto-expanded block) to a terminal and advances the cursor. Pressing it repeatedly steps through a file, executing block by block.
- **`SendLine`** — sends the current line and advances. If text is selected, sends the selection.
- **`SendFile`** — sources the entire file using language-specific commands (`source()` for R, `exec(open().read())` for Python, `include()` for Julia).
- **Blank lines** send a bare Enter, so stepping through code feels natural.

## Language-aware block expansion

This was the big ask from @polymorphicengine and @casperhart — not just sending selected text, but intelligently selecting the right block based on cursor position:

- **Python**: Expands to the enclosing function/class body (indentation-based), supports `# %%` Jupytext cells, includes decorators
- **R**: Follows pipe chains (`%>%`, `|>`), ggplot `+` continuations, bracket balancing across multiline calls
- **Julia**: Uses tree-sitter outline queries, falls back to keyword/end pair matching
- **All other languages** (shell, text, etc.): Falls back to single-line behavior — no over-eager paragraph grabbing

Comments are handled intelligently — standalone comments are boundaries between expressions, but comments embedded inside bracketed expressions (e.g., inside `inner_join(...)` in R) are included as part of the block.

## Multiple send targets

Sends to Zed's built-in terminal panel by default, but also supports external terminals via AppleScript/CLI:

- **Ghostty**, **iTerm**, **Terminal.app** (AppleScript)
- **tmux**, **cmux** (CLI subprocess)

Per-language target routing lets you send R to Ghostty, Python to a tmux pane, etc.

## Bracketed paste

Multiline code is wrapped in bracketed paste escape sequences so REPLs receive it as a single paste rather than executing line-by-line. For targets/languages where this doesn't work well, it falls back to line-by-line sending with small delays.

## Debug mode

Set `"debug": true` in settings to get a confirmation toast before each send, showing the detected language, target, and a preview of the extracted text. Click "Send" to confirm or dismiss to cancel.

## Addresses specific requests from this thread

- @JosiahParry's request for `cmd+enter` to execute code in a REPL — this is exactly what `SendCode`/`SendLine` do with user-configured keybindings
- @polymorphicengine's request for sending a "block" of code (lines without free lines in between) — `SendCode` does this with language-aware expansion that's smarter than just paragraph detection
- @casperhart's request for inferring the appropriate function/class/statement based on cursor position — `SendCode` does this for Python, R, and Julia via indentation analysis, pipe/bracket continuation, and tree-sitter outlines
- @ddxv's workflow of writing code in the editor and testing in an interpreter below — this is the core use case
- @gsamat's reference to VS Code's `RunSelectedText` — `SendLine` with a selection does this, and `SendCode` goes further with block expansion
- The `SendKeystrokes` workaround from @ToddMCr and @mcabbott works but clobbers the clipboard and can't do block expansion — `SendCode` writes directly to the terminal without touching the clipboard

## Setup

```json
// keymap.json
[
  {
    "context": "Editor && mode == full",
    "bindings": {
      "cmd-enter": "send_code::SendLine",
      "ctrl-enter": "send_code::SendCode",
      "shift-cmd-enter": "send_code::SendFile"
    }
  }
]
```

```jsonc
// settings.json
{
  "send_code": {
    "target": "zed_terminal",
    "bracketed_paste": true,
    "language_targets": {
      "R": "ghostty",
      "Python": "cmux"
    }
  }
}
```

## Why a core crate (not an extension)

This is implemented as a core crate because it needs editor selection access, terminal panel integration, and custom action registration — none of which are available in the WASM extension API.

Code is at `crates/send_code/` in [my fork](https://github.com/scfurl/zed). The Zed terminal sender is well-tested and working. The external terminal senders (Ghostty, iTerm, Terminal.app, cmux, tmux) are implemented but have not been fully tested yet — feedback and bug reports on those are welcome.

Happy to hear feedback on what else would make this useful for your workflows.

**Reply by JosiahParry** [reply]
@scfurl please see https://github.com/zed-industries/zed/pull/42467
